### ISSUE TITLE
Support fractional abundance rates for thermodynamic equilibrium

### ISSUE THREAD

**Issue by mattngc** [cherab/openadas#16]
This is a desired new feature. @Mateasek has already contributed some code in #4. To address this issue we will need to create a number of new features.

**Comment by Mateasek** [comment]
To have a small update after all the canges to the ADF11 atomic rates and interpolation. I uploaded new version of the outline for the balance calculation [here](https://github.com/Mateasek/cherab_openadas/blob/feature/frac_abundance/demos/frac_abundance.py).

I added the possibility to include thermal cx into the balance equation. The results of are:
![ionbal_v2](https://user-images.githubusercontent.com/969455/55388720-0f9b3100-5534-11e9-9459-68d7efde906e.png)

The result now seems more realistic. There are some problems with steep changes in regions where one of the rates gets low and the drate/dte is high. I think that is the limit of openadas data. What is the problem tho is that the np.linalg.lstsq used for solving the matrix actually allows negative results. That is bad. I will have to test some other function for example from scipy.

Apart from this I think the openadas data are ready for ionization balance calculations.

**Comment by Mateasek** [comment]
I have utilized the scipy.optimize.linear_lsq function to solve the balance matrix [here](https://github.com/Mateasek/cherab_openadas/blob/feature/frac_abundance/demos/frac_abundance.py). It allows to set the bounds for the solutions which can be used to prevent getting negative densities. Also the balance solution is now obtained using the unnormalized matrix, i.e. the solutions sum up to electron density. For some reason all the optimization functions behave better when they are working with larger numbers or I did something wrong. To have comparison with the previous state I'm uploading the ionzation balance for neon again.

![ionbal_compare](https://user-images.githubusercontent.com/969455/56300514-64df6100-6136-11e9-9f92-0d8e2a05a230.png)


I think the example is now finished and the calculation now is complete. It would be nice if @mattngc could compare the results with ADAS results, so we see if we get results with acceptable agreement.

**Comment by mattngc** [comment]
Excellent, really nice work @Mateasek. I'll see if I can generate the equivalent case in ADAS for us to benchmark against.

**Comment by mattngc** [comment]
I've performed a benchmark against the run_adas405() routine in the ADAS subscription package. The calculations were performed for Nitrogen at a density of 1e19 m^-3. I have made two plots of the abundances, with ADAS official in solid lines and @Mateasek's  contributed code in dashes.

![N_frac_abun_loglog](https://user-images.githubusercontent.com/8628876/57086843-d0106200-6cf6-11e9-932d-2efd71e916fe.png)

![N_frac_abun_loglin](https://user-images.githubusercontent.com/8628876/57086862-d69ed980-6cf6-11e9-9dda-05e49d6fce6b.png)

The discrepancies are barely visible. So if you open a pull request for the demo I'll merge it into the OpenADAS demos and we can close this ticket.
